modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/SortedExternalResult.java

```diff
@@ -0,0 +1,473 @@
+        for (Value[] row : rows)
+            addRow(row);
+
+        return size;
+    }
+
+    /** {@inheritDoc} */
+    @Override public int addRow(Value[] row) {
+        if (isAnyDistinct()) {
+            if (containsRowWithOrderCheck(row))
+                return size;
+        }
+
+        addRowToBuffer(row);
+
+        if (needToSpill())
+            spillRowsBufferToDisk();
+
+        return size++;
+    }
+
+    /**
+     * Checks if current result contains given row with sort order check.
+     *
+     * @param row Row.
+     * @return {@code True} if current result does not contain th given row.
+     */
+    private boolean containsRowWithOrderCheck(Value[] row) {
+        Value[] previous = getPreviousRow(row);
+
+        if (previous == null)
+            return false;
+
+        if (sort != null && sort.compare(previous, row) > 0) {
+            removeRow(row); // It is need to replace old row with a new one because of sort order.
+
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns the previous row.
+     *
+     * @param row Row.
+     * @return Previous row.
+     */
+    @Nullable private Value[] getPreviousRow(Value[] row) {
+        assert unsortedRowsBuf == null;
+
+        ValueRow distKey = getRowKey(row);
+
+        Value[] previous;
+
+        // Check in memory - it might not has been spilled yet.
+        if (sortedRowsBuf != null) {
+            previous = sortedRowsBuf.get(distKey);
+
+            if (previous != null)
+                return previous;
+        }
+
+        // Check on-disk
+        Long addr = hashIdx.get(distKey);
+
+        if (addr == null)
+            return null;
+
+        previous = readRowFromFile(addr);
+
+        return previous;
+    }
+
+    /** {@inheritDoc} */
+    @Override public int removeRow(Value[] values) {
+        ValueRow key = getRowKey(values);
+
+        if (sortedRowsBuf != null) {
+            Object prev = sortedRowsBuf.remove(key);
+
+            if (prev != null)
+                return size--;
+        }
+
+        // Check on-disk
+        Long addr = hashIdx.remove(key);
+
+        if (addr == null)
+            return size;
+
+        Value[] res = readRowFromFile(addr);
+
+        if (res == null)
+            return size;
+
+        markRowRemoved(addr);
+
+        return size--;
+    }
+
+    /**
+     * @return {@code True} if it is need to spill rows to disk.
+     */
+    private boolean needToSpill() {
+        return !memTracker.reserved(0);
+    }
+
+    /**
+     * Adds row in-memory row buffer.
+     * @param row Row.
+     */
+    private void addRowToBuffer(Value[] row) {
+        long delta = H2Utils.calculateMemoryDelta(null, null, row);
+
+        memTracker.reserved(delta);
+
+        if (isAnyDistinct()) {
+            assert unsortedRowsBuf == null;
+
+            if (sortedRowsBuf == null)
+                sortedRowsBuf = new TreeMap<>(cmp);
```

Comment:
Clarifying question: why do we use `TreeMap` here? Is it only in order to provide custom `Comparator`? I think about renaming `sortedRowsBuf` to `mapRowsBuf` if a sorting order of the map is not used.